feat(audio): add real gain control of the input device

antis81 opened the pull request:
This PR adds a long awaited feature. We now have a controllable **Audio Gain**, no really! :smile:

Gain can be controlled in positive/negative dB values from -30~30dB and allows for amplification or mitigation of the input signal level, resulting in highly improved audio quality. Thanks to @initramfs for the technical advice and review.

~~The gain is calculated with the common `10*log10()` algorithm, which is normally used for dB calculation of the electrical power level - and it gives pretty good results. Values can be set via qTox's public Audio API as percentage values (between 0 and 1).~~

Note: This PR also brings back const-correctness to `Audio`; some minor cleanup with naming conventions also included.

~~Important: As this has quite an impact on qTox's Audio experience --> Anyone who is able to compile qTox is invited to try this PR from a users view point!~~

The following values should fit for most common microphones, producing splendid clear results. Helps to check vs. [Tox EchoBot](https://toxme.io/u/echobot)):
- System settings:
  - Master Volume: ~80%
  - Gain: ~50% (~10dB)
- qTox Audio Settings:
  - Master volume: ~80%
  - Gain: ~60-85%

**TODO:**
- [x] Add sane min/max gain boundaries (-30/+30 dB) in `Audio`.

**UI will be done in separate PR**
- Add tick marker in slider at "unity gain" level.
- Add clipping indicator.
- [x] ~~Change label text in A/V settings to "Input gain" or only "Gain".~~ --> Wrong German translation.

Closes the following issues:
- close #3109 
- close #2756 

Have fun talking with qTox!

zetok commented:
cc @towlie ?

towlie commented:
For what am I needed here? Only compiling for windows?
I saw no issues while testing.

antis81 commented:
> For what am I needed here? Only compiling for windows?

@towlie This and I'd like to get some feedback about how usable that is. Yep, thanks!

If someone could check it shortly on OSX, this would be nice. Other than that -> we're ready to merge.

initramfs commented on src/audio/audio.cpp:
```diff
@@ -486,16 +498,30 @@ void Audio::doCapture()
+    for (quint32 i = 0; i < AUDIO_FRAME_SAMPLE_COUNT * AUDIO_CHANNELS; ++i)
+    {
+        int16_t ampPCM = static_cast<int16_t>(qRound(buf[i] * gain));
+        buf[i] = static_cast<int16_t>(ampPCM);
```
Why is the `int16_t` being cast again to an `int16_t`?

antis81 replied:
~~Indeed not required here. Will remove it.~~

Fixed!

initramfs commented on src/audio/audio.cpp:
```diff
@@ -486,16 +498,30 @@ void Audio::doCapture()
-        buf[i] *= inGain;
+    for (quint32 i = 0; i < AUDIO_FRAME_SAMPLE_COUNT * AUDIO_CHANNELS; ++i)
+    {
+        int16_t ampPCM = static_cast<int16_t>(qRound(buf[i] * gain));
```
Given that the gain can be as high as 20 now, this line does not stop the samples from railing against `INT16_MAX` (or it's negative), insteading allowing it to roll over, creating horrible distortion instead of digital peaking.

antis81 replied:
Fixed!

initramfs commented:
This code dramatically changes the gain characteristics of the input source from having unity gain at 100% to having unity gain at 10^(1/10)% (~ 1.26%). Whilst this would work well for most microphone sources who's preamps supply signals way below the input line level, people who have microphones or input sources that supply at nominal line levels will have a very hard time controlling the gain since deviation in either direction (around ~1.26%) results in a very large change in the resulting gain.

If possible, I would prefer if unity was somewhere at say 30% of the way (or at worst 10% of the way) with a small marker indicating where it is (and perhaps the slider can snap to it when near).

initramfs commented on src/audio/audio.cpp:
```diff
@@ -486,16 +498,30 @@ void Audio::doCapture()
+    percentage = std::max(0.0, std::min(percentage, 1.0));
+    gainPercentage = percentage;
+
+    gain = (percentage > 0.0) ? 10.0 * qAbs(log10(percentage * 100.0)) : 0.0;
```
The use of `qAbs()` in this context causes incorrect behaviour when input gain percentage is < 1% (it causes a massive rise in gain since the log curve is reflected upwards).

antis81 replied:
Fixed!

initramfs replied:
Removing the abs is no better since now gain can go negative (causing channel inversion) with the same effect (increasing negative gain). If this new gain curve is to be used, the log value has to be clipped at some lower bound (0 for example, such that it doesn't go negative).

antis81 replied:
Yeah, I just remembered why I put it there. Actually removing the abs changes "nothing" here as values are always within range 1..100 (percentage \* 100.0) -> I'll re-add the abs to be on the safe side. Btw. Percentage is already clipped above.

Still 1% results in a 0 gain factor…??

initramfs replied:
Sorry, maybe I'm missing something but where does it indicate that percentages are always greater than 1%? If this is guaranteed by some other code perhaps add a comment somewhere (in the function documentation perhaps). You may need to check consistentcy elsewhere as well (particularly in `inputGainPercentage()` where the return documentation states percentage can be anything between 0 and 100%).

antis81 commented:
> If possible, I would prefer if unity was somewhere at say 30% of the way (or at worst 10% of the way) with a small marker indicating where it is (and perhaps the slider can snap to it when near).

Sure, the markers are fine. We don't want to snap, as it's less accurate and resolution is kinda rough by 1% per tick. So let's see how the markers work for users. I'd rather finish and re-animate the gain level meter providing a visual feedback.

@initramfs Thanks for the discussion. Really helpful. :)

TheNain38 commented:
@antis81 All of the functions and variables having `percentage` in their name in this code, aren't actual percentage, normally, this is a value between 0 and 100 (because it is a value to be divised by 100), so, I suggest you rename these to something else than `percentage`, because it is inducing that this should be a value between 0 and 100.
See [here](https://en.wikipedia.org/wiki/Percentage)

antis81 commented:
@TheNain38 Thanks for the comment. So what? :)

Like written in the API description, the gain level describes a non-linear amplification on a logarithmic scale (0..20dB). Any linear representation (e.g. in percent) is by far not optimum. A gain level of 50% amplifies the input signal already by ~17dB, while 100% represents maximum 20dB amplification. The optimum would be probably to actually create a logarithmic slider, which we currently do not have. :)

TheNain38 commented:
@antis81 I mean, in your code, for example, [here](https://github.com/tux3/qTox/pull/3311/files#diff-f23c32cfb93eaa31ca828e9e9f75b2a8R517), you called the argument `percentage`, when it actually is a value between 0 and 1, and not a percentage (which is between 0 and 100 by definition), so I was suggesting you just change the name of variables and functions, using `percentage` in their names (whithout it being a real percentage) :)

initramfs commented:
@TheNain38 To my knowledge a percentage essentially represents a fraction out of 100. The common notation of appending the % symbol after a valid indicates the value is to be taken (divided) out of 100. In the wikipedia article you linked (emphasis mine):

> For example, 45% (read as "forty-five percent") is equal to 45⁄100, or **0.45**.

In my experience I've seen percentages expressed (in a programmatic context) as both integers (0 ~ 100) and floating points (where 1.0f is 100%). As such, I don't see a problem with the representation antis81 choses to use within this PR.

I may have misrepresented your concern in terms of consistentcy. If percentages are represented elsewhere as `qreal`s bounded between 0.0 ~ 100.0 (especially within Qt's libraries) then I support your decision to change public functions to take inputs bounded between 0.0 ~ 100.0.

Just my two cents.

initramfs commented:
@antis81 You need to be careful with your terminology, this code **does not** provide 20 dB of gain, it instead provides around 26.02 dB of gain. Absolute gain value is **not** equal to decibel gain value. At 50% of the input "gain percentage" the actual gain is ~24.6 dB.

Upon further inspection of your code, it appears you may have the volume control backwards (i.e. you need to use exponentials instead of logarithms). Since human hearing scales exponentially (i.e. you need exponentially more power/voltage/signal for a linear increase in perceived volume).

Take a look at [this diagram](http://sound.westhost.com/pots-f4.gif) denoting graphs of potentiometers (passive electronic devices that are commonly used as volume knobs for audio appliances). The "log pot" is somewhat of a misnomer in this regard.

Furthermore, I think the audio code would benefit greatly if all percentage units are replaced with dB units since percentage units are extremely vague.

As it currently stands, when this PR gets merged existing percentage units saved in configuration files will now mean something completely different. If I had my input gain/volume (this PR makes this terminology vague) set to 30% this would translate to a 70% **reduction** in absolute signal level (or a 10.45 dB **reduction**). After this PR gets merged the same slider value of 30% will now translate to a 1377% **increase** in absolute signal level (or a 23.39 dB **increase**).

If the audio code could instead simply deal in linear units of dB (and calculate the gain based on that). It would make things much simpler (and allow my earlier concern regarding unity gain at 30% to be implemented more easily).

See [this page](http://www.sengpielaudio.com/calculator-amplification.htm) on calculating dB gain from absolute gain.

antis81 commented:
@initramfs 
Glad talking to someone interested in the maths behind the gain topic. I'm just scratching the surface (with the objective to make things better for the qTox community). :)

**FYI:** With the algo and value range used in this PR we get **only increased** PCM values. The minimum factor `10*log(2) = 3.01". Value of 0 or 1 results in a muted signal, which is unwanted here of course -> we want the factor to be 1 (==no amplification) in this case. Thus, there's never a decrease happening as the factor doesn't drop below 1.

:question_mark: Does decreasing the input level make any sense in our use case?

After thinking and reading more about adding gain to PCM encoded signal, I think we are best off by providing several lookup tables with the pre-calculated factors in a defined resolution. Current resolution would be 100 values, but we can use 200 or 500 as well. The advantage of this is, that we can set the slider's maximum to `sizeof(arr)-1` and it always uses the correct resolution. Further we can actually map the distance to linear values, so that a slider in the middle represents actually "50%" of maximum amplification. For the algo I think it's best to let the user decide between logarithmic, exponential or unity gain - depending on what fits best for him/her. For me, the `10*log10` gain provides excellent results and I do not need anything else.

I'd like to state out one thing clearly:
I'm always talking about **input gain**, not "input volume", which absolutely makes no sense in this context. Maybe this came out a little confusing in the code due to "historical reasons", but gain is what I always meant in this context.

This being said, I don't think the actual dB does matter to qTox users and we're good to go without exposing any technical background.

> Furthermore, I think the audio code would benefit greatly if all percentage units are replaced with dB units since percentage units are extremely vague.

Well, we should use the right thing for the right purpose. So… percentage for output volume (OpenAL and Qt both use the 0..1 percentage terminology) and dB (I think it's best using a LUT with index) for the gain.

@initramfs Would you mind creating the LUT(s) with maybe 100 (or maybe better resolution) `float` factors?

initramfs commented:
@antis81 

> With the algo and value range used in this PR we get only increased PCM values.

Which, under the conditions of uniform gain application applied across all samples will translate into a output gain of the same factor.

> Does decreasing the input level make any sense in our use case?

Yes, because not everyone uses input devices that input at low levels. Some people may have devices that have a high average level (of say - 3 dBFS) and want to cut it down to match their friends for example. I understand the fact that lowering the gain is a feature that'll be less common for the masses of people using qTox but that's the current way that input is being handled (see below), merging this PR would essentially remove this feature (which I personally find useful anyway).

> After thinking and reading more about adding gain to PCM encoded signal, I think we are best off by providing several lookup tables with the pre-calculated factors in a defined resolution.

By using decibel measure directly, one can linearly interpolate without loss the appropriate gain value for a given percentage/slider position, eliminating the need for any lookup table (see below).

> For me, the `10*log10` gain provides excellent results and I do not need anything else.

This is probably due to the fact that the log curve rises very quickly initially, allowing good control of gain that somewhat seems reasonable. Whilst this is not a bad idea, I personally believe you should give a try to my proposal below.

<hr>

Just to clear things up, I'll state a new proposed algorithm for gain control (it's not really new, the same algorithm used in basically all gain control systems that I've dealt with [I'm somewhat of an audio engineer]).

The current qTox implementation of "gain" is best described like this:
![master-gain](https://cloud.githubusercontent.com/assets/3128050/15455865/e4e7521e-2092-11e6-8998-fe20d14fe00c.png)

The y axis "Gain (absolute)" represents the actual gain multiplier applied to the PCM samples whereas the x axis "Gain (percentage)" represents the input percentage taken by `audio.cpp` as well as the slider position. In this (and all futher diagrams) the blue line represents the relation and the red line represents the position of unity gain.

In the current qTox implementation of input "gain", we have something that's essentially a linear cut in which 100% represents unity and 0% represents 0 absolute gain. In this system the highest gain is +0 dB (i.e. no amplification at all) whilst providing a linear cut to -∞ dB. It's clear the see why people will complain about low microphone levels since there is no way to digitally boost the levels beyond what the line in level from the audio source is.

Your PR changes the gain characteristics to this:

![pr-gain](https://cloud.githubusercontent.com/assets/3128050/15455923/214bdd96-2094-11e6-825d-0db38d17bfb4.png)

It's easy to see why this PR can satisfy the needs of most since it massively increases the digital gain to ~26.02 dB (instead of 0.0 dB from the current implementation). Not only that, the log curve fits nicely since it boosts gain very fast as you traverse up the slider which makes it sound like a linear travesal in slider position is indeed causing a proportional increase in perceived audio output. The shortfalls of this method are that it no longer provides a way to accurately control gain values below 0.0 dB. Another issue is that for people with truly quiet input devices that require 80% or more gain, the changes between 80, 90 and 100% sliders don't feel like they change audio output level at all (given the nature of the curve compounded with the nature of human hearing).

My proposed method of control is to treat the input not as percentages, but directly as decibel values. The slider can have something like -20 dB on one side and +20 dB on the other in which the slider will **linearly** traverse between the maxima and the minima. This is illustrated as:

![proposal-20](https://cloud.githubusercontent.com/assets/3128050/15455980/8420f748-2095-11e6-9b10-ba7ec142b6ec.png)

Note how using decibel measure (and then converting it to absolute gain via exponents) allows a linear value (dB/slider position) to be transformed into gain values that scale with human hearing. Not only that, such a control allows both amplification and damping to be applied to the PCM samples, retaining existing functionality whilst adding the ability to boost the samples by large gain ratios. In the case that people see more value in amplfication than in damping, we can bias the 0 dB point more to the left, allowing more positive gain headroom for amplication (as shown below):

![proposal-10-30](https://cloud.githubusercontent.com/assets/3128050/15455978/7fc3989a-2095-11e6-8cb2-9abd1a3b4d1a.png)

The choices of values (20 dB/-10~30 dB) is arbitary here, any range of values will work (I chose said values to best match the range that you've set out and had success with). No LUT is required since we can map any slider positon/value to a decibel value via simple linear algebra. Any precision can be specified by the slider since the dB -> gain multiplier algorithm is a simple exponential (more precisely `10^(dB / 20)`) which can be calculated as stored everytime the slider is moved (as your code currently does).

The user does not need knowledge of what decibels is, only that a linear change in the slider results in a perceived linear increase/decease in output level throughout the whole slider (e.g. moving from 50% of the slider to 60% of the slider feels like it gets louder by the same quantity as moving from 80% to 90%).

I urge you try this algorithm and see if it works for you (I really don't see why it won't).

antis81 commented:
@initramfs Great to hear you are kind of involved in sound design - I'm kind of hobbyist musician :). I implemented your proposed algorithm. **Surprise:** With my lousy setup, the log algorithm works better. With this information, I think the optimum would be combining the linear algo for negative amp and the log for positive amp. With the 10^(dB/20) algo I found, that a range from -20dB~12dB (x10 => 320 slider ticks) fits best. This is wide range and values near or above 12dB easily result in heavy distortion and clipping effects. So the proposed algo is a good trade-off, but not optimum. I will add the tick marker at unity gain (see todo list in description). After this is done we're pretty much through with the PR and ready to merge.

initramfs commented:
@antis81

> This is wide range and values near or above 12dB easily result in heavy distortion and clipping effects.

That cannot be right, 12 dB corresponds to an absolute gain of 10^(12/20) = ~3.981. From your initial comments, you claim that you get good results with your log algorithm at the 50~65% range, this corresponds to an absolute gain of **at least** 16.89 or nearly **4.27** times larger than what you claim to cause distortion. Are you sure you've implemented the proposed algorithm correctly?

> I think the optimum would be combining the linear algo for negative amp and the log for positive amp

I would oppose changing algorithms between positive and negative gain since it essentially causes confusion as the sliders no longer change sound in any way proportional to the perceived audio change.

> I found, that a range from -20dB~12dB

Any reason to put more on the negative side? I thought you brought up the point that negative gain isn't very useful (which I somewhat agree with and thus proposed a bias towards positive gain).

antis81 commented:
> Are you sure you've implemented the proposed algorithm correctly?

Yep. But maybe we are misunderstanding. Please have a look for yourself in `Audio::privateSetGain` at [audio.cpp:L518](https://github.com/tux3/qTox/pull/3311/commits/a558f03967ea4a77a4b5745cb5b28de2b8224c7a#diff-f23c32cfb93eaa31ca828e9e9f75b2a8R518).

initramfs commented:
@antis81 Yep, that implementation is correct.

If possible, can you print the associated gain dB vs gain absolute values for +12 dB versus the gain values you get for your log curve/algorithm at 50% percent position? Because my earlier concern still holds...

[audio.cpp:L262](https://github.com/tux3/qTox/pull/3311/files#diff-f23c32cfb93eaa31ca828e9e9f75b2a8R262) uses an incorrect multiplier to calculate dB gain values from settings.

antis81 commented on src/audio/audio.cpp:
```diff
@@ -251,6 +259,8 @@ bool Audio::initInput(QString inDevDescr)
         return false;
     }
 
+    privateSetGain(Settings::getInstance().getInVolume() * 0.01);
```
Of course multiplier no longer applies here -> will be removed. Thanks @initramfs.

antis81 commented on src/audio/audio.cpp:
```diff
@@ -486,16 +496,32 @@ void Audio::doCapture()
+void Audio::privateSetGain(qreal dB)
+{
+    gainDecibel = dB;
+    gain = qPow(10.0, (dB * 0.2));
```
@initramfs Duh! Factor is wrong -> will be changed to **0.05**. Will add some graphs visualizing the (corrected) values. Stay tuned.

initramfs replied:
@antis81 Wow... I missed that too... Must have been real late when I read that.

antis81 commented:
@initramfs 

Here's the graph for my first logarithmic approach with positive (percentage) integer values ranging from 1~100:
![logarithmic](https://cloud.githubusercontent.com/assets/440517/15487435/a268e0fe-214d-11e6-8c05-b597555ca441.png)

And here's current (corrected) exponential approach using lower amp factors within ~~current~~ range -20~12dB:
![exponential](https://cloud.githubusercontent.com/assets/440517/15487525/4e304a30-214e-11e6-84fc-4a778b2c2669.png)

I now widened the range from **-30~30dB**, which now also provides good results in both directions (lowest factor: ~0.03 ; highest factor: ~31.62).

initramfs commented on src/audio/audio.cpp:
```diff
@@ -251,6 +259,8 @@ bool Audio::initInput(QString inDevDescr)
         return false;
     }
 
+    privateSetGain(Settings::getInstance().getInVolume());
```
Just to tidy up, you need to make sure the input values from the settings are sane (since users with old configurations may contain some very wacky values). I'd suggest defaulting to unity gain (0.0 dB) if values are out of range.

antis81 replied:
Yes, that's totally ok. I guess it makes sense adding a (probably customizable) `minGain/maxGain` values to Audio API, defaulting to +-30dB.

initramfs commented:
@antis81 Read the line comment I made earlier, but everything else looks good to me. A few small (but not entirely necessary) things that can be added are a marker to where unity gain is (halfway on the slider), a peak indicator (if the UI allows for it) to indicate when samples are railing against 0 dBFS.

antis81 commented:
@initramfs We had a gain meter implemented (removed it, because it was unfinished) and I think it makes sense reviving the idea, only doing it right this time.

Both, the marker and a gain indicator or meter are totally doable with acceptable effort.

antis81 commented:
@initramfs
Could you do a last complete code review please? I am going to implement the UI features in another PR as this introduces new widget(s) completely independent of qTox audio API.

Most relevant changes:
- added the min/max gain values; set to -30/+30dB; stored in `Audio::Private` object
- min/max gain corrects wrong user settings; **defaults** to 0dB unity gain, if settings value does not exist; settings value **bounds** to min/max gain in dB (e.g. previously 100% ==> +30dB ==> users will moan about great feedback for one last time :smile:).
- audio input device is initalized with gain from settings ("audio/inVolume") in `Audio::initInput`; same applies to slider in a/v settings form; **result:** Users will no longer think, that settings are not correctly saved, because slider always jumps to minimum/maximum. :)

**One important thing I noticed when implementing this:**
As the term "inVolume" is wrong in settings anyway, IMO it is best to introduce a new value "audio/inGain" and leave the old as a "dead" value. This would reset everything for users and initialize with the 0dB unity gain. Would also prevent people from moaning about the big feedback, before checking their audio configuration.

@tux3 @zetok
Whoever writes the documentation: We should give a hint to users in release notes, that audio settings changed and they need to re-adjust their in/out levels (setting sliders to out:~80/in:~70 "percent" is a good common setup).

antis81 commented:
@TheNain38 Btw.: No percentage values were harmed during the making of this PR (**EDIT:** a.k.a no percentages are used any longer). :smile:

initramfs commented on src/audio/audio.cpp:
```diff
@@ -36,6 +37,37 @@
+    void setInputGain(qreal dB)
+    {
+        gainDecibel = qBound(minInputGain, dB, maxInputGain);
+        gain = qPow(10.0, (dB * 0.05));
```
The parameter `gain` is being calculated from the unbounded input argument `dB`, resulting in incorrect behaviour when out of bounds values are used (i.e. whilst `gainDecibel` will correctly clip to +-30 dB, `gain` will be arbitarily set based on the input dB value, which could potentially come from wrong/outdated settings).

Not an issue but, is there a particular reason to use `dB * 0.05` instead of `dB / 20.0`? I feel the latter is more clear.

antis81 replied:
Thanks, will correct that.

As we have a constant value, there's no reason. When factor is variable, this avoids "division by zero" error.

initramfs commented on src/audio/audio.cpp:
```diff
@@ -36,6 +37,37 @@
+        gain = qPow(10.0, (dB * 0.05));
+    }
+
+public:
```
Keep these fields public somewhat defeats the above function. Any function can arbitarily modify the fields without the logic defined above. In particular, it allows for the consistentcy of the four parameters to be broken by mistakes that are no easily seen (since they can be anywhere in the code base that uses this code). I understand that mutation is required by the `Audio` class, see my line comments below.

antis81 replied:
> Keep these fields public somewhat defeats the above function. …

Yep, we can make them immutable by adding one or two more wrapper functions to completely control this throughout qTox Audio API.

initramfs commented on src/audio/audio.cpp:
```diff
@@ -125,28 +158,78 @@ Set the master output volume.
-    inGain = std::max(0.f, std::min(volume, 1.f));
+@note Default is 30dB; usually you don't need to alter this value.
+*/
+void Audio::setMaxInputGain(qreal dB)
```
What is the advantage of creating a whole new class to encapsulate mutable parameters that that are essentially only changed by this class? I feel like that makes the code less readable by the constant access to object `d`. Isn't holding the values as member fields of this class good enough? Comments and variable isolation can help with establishing that the values +-30 dB are defaults. You can also declare the defaults as `constexpr` and completely make them immutable, allowing modification only via code changes.

If the separate class approach is necessary in your opinion, please move the `setX` `getX` functions that relate to gain to that class, or else the `Audio::Private` class is merely a glorified `struct` whose functionality merely adds an additional layer of indirection to otherwise things that could have been member fields of `Audio`.

Another potential idea to reduce the size of the `Audio` class would be to declare a `AudioGainDevice` (or a name of similar sort) that defines an abstract/virtual class that provides functions for gain control with suitable defaults. The `Audio` class can then appropriately derive from it as a subclass. Access control can be done via private/protected fields in the base class.

antis81 replied:
> What is the advantage of creating a whole new class to encapsulate mutable parameters that that are essentially only changed by this class?

It is more than just that. Actually, I prepare to encapsulate the qTox Audio API completely and bring back the pimpl idiom. This is absolutely necessary, to keep the API controllable, making it "future proof" (had a hard time btw. collecting all the code fragments and clean up the mess to get to the current stable state). Further it actually eases handling of audio data within qTox by providing a sane interface, which cannot be mistaken. It has some other advantages as well. For example it allows an atomic replacement of underlying `OpenAL` with another library (e.g. RtAudio).

> You can also declare the defaults as constexpr and completely make them immutable, allowing modification only via code changes.

This also was my first thought, but it is just not the right thing to do here. Those values are really meant as defaults and thus mutable through public audio api.

> Another potential idea to reduce the size of the Audio class would be to declare a `AudioGainDevice`…

Yes, why not. I'd simply name that class `qTox::Audio::InputDevice` (including the namespaces). I think about splitting up audio devices for some time now, being able to actually use them when required and replace the "crappy" singleton and subscription mechanisms. OTOH, this PR is actually ready, making qTox "audio conferencing" finally controllable to users, which probably helps to extend the community (user base). I don't want to break that by bringing in new concepts too quickly. Hope you understand. Anyway I'd love to discuss future of qTox Audio API with a "professional attitude". ^^

initramfs commented on src/widget/form/settings/avform.cpp:
```diff
@@ -65,6 +67,12 @@ AVForm::AVForm() :
+
+    bodyUI->microphoneSlider->setToolTip(
+                tr("Use slider to set the gain of your input device ranging"
+                   " from -30dB to +30dB."));
```
These numbers should be retrieved from the defined constants currently in `Audio::Private`

antis81 replied:
Yep, thanks. Will change that.

initramfs commented:
> As the term "inVolume" is wrong in settings anyway, IMO it is best to introduce a new value "audio/inGain" and leave the old as a "dead" value.

I fully support this idea, in fact I was thinking about it since I first saw what this PR did since it radically changes the way input is handled (it is no longer a linear volume, but an actual digital amplifier now). Moreover, we can derive input gain from the existing input volume field via the equation `20 * log10( input_volume)` such that users who eventually run a version of qTox with this PR merged will not hear any difference initially (they'll only notice that the slider is now in a different position [much lower than where they initially placed it before this PR, allowing for a lot more extra headroom]).

antis81 commented:
> I fully support this idea, in fact I was thinking about it since I first saw what this PR…

Ok, I'll do that as well then and rename the settings value.

antis81 commented:
@initramfs The discussed changes are ready. Merge?

antis81 commented on src/audio/audio.cpp:
```diff
@@ -36,6 +37,47 @@
+    void setInputGain(qreal dB)
+    {
+        gain = qBound(minInputGain, dB, maxInputGain);
+        gainFactor = qPow(10.0, (gain / 20.0));
```
@initramfs Hint: I exchanged meaning to `gain` (in decibel) and `gainFactor`, so it is more clear.

antis81 commented on src/audio/audio.cpp:
```diff
@@ -36,6 +37,47 @@
+    qreal   minInputGain;
+    qreal   maxInputGain;
+
+private:
```
Look, a private private -> neat. :) Now values can only be changed through `setGain()`, making it "fool proof".

antis81 commented:
@initramfs A small "post implementation UI change": I added tickmarks to the gain slider, improving usability. Unfortunately QSlider does not provide setting tickmarks on an array of certain positions, but on a fixed interval only. This requires implementing our own slider widget (started with implementation, but it will take some time).

antis81 committed:
feat(audio): add slider tickmarks, improving better visible orientation

initramfs commented:
@antis81 

> It is more than just that. Actually, I prepare to encapsulate the qTox Audio API completely and bring back the pimpl idiom.
> 
> I think about splitting up audio devices for some time now, being able to actually use them when required and replace the "crappy" singleton and subscription mechanisms.

Well okay, then I guess I can see where this is going. Forgive my earlier comments on this matter then since I didn't know exactly where this was leading to. Funnily enough, I've been doing the same thing for qTox's video code (replacing singleton patterns and allowing multiple object/devices). In particular, my changes plan to finally allow for proper desktop sharing control.

Regarding the settings change from `inVolume` to `inGain`, I noticed you haven't added any conversion code to transform old `inVolume` settings to `inGain`, is this intentional? I mean, it's not really that much of a big deal, I just want to clarify.

Apart from that there is just a few stylistic changes that is required (some opening brackets need to be on new lines). Also, can the private instance be named something more descriptive than `d`?

antis81 commented:
> I noticed you haven't added any conversion code to transform old inVolume settings to inGain, is this intentional?

Yes. Like discussed, the `inGain` value is "reset" to unity gain (this matches the before 100% `inVolume`). This will be least painful for users and exceeds the range seamlessly. Converting the old linear values to negative exponential dB values from -30~0dB doesn't pay off, don't you think? Btw. apparently qTox drops the old `inVolume` settings key - that's a good thing actually.

> Also, can the private instance be named something more descriptive than d?

The `d` nomenclature is a quite common practice and therefore actually very descriptive. It is even well documented in Qt's internal and public docs and used consequently throughout the library, where I adopted some quite useful programming techniques from. :) For an example, have a look at `QExplicitlySharedDataPointer` and `QSharedData`. This is actually what I have in mind for the future to split audio devices. I started a "test implementation" locally to see how much effort this will become. In short, this is unfinished and it will take several days to complete.

antis81 commented:
> Apart from that there is just a few stylistic changes that is required…

Whoopsy, will be corrected. I'd like to merge afterwards keeping focused on what's most relevant here. Could you review #3309 and #3175 as well? The latter is another candidate for moving the abstraction layer to the `d` pointer. And now it starts to create a more complete picture… :)

antis81 committed:
feat(audio): add real gain control of the input device

The gain can be set now in dB values, providing acceptable results for a wide range of setups.

note: also introduces const-correctness and some minor cleanup.

initramfs commented:
> Converting the old linear values to negative exponential dB values from -30~0dB doesn't pay off, don't you think?

Yep, no problem, just wanted to verify correct behaviour.

> The `d` nomenclature is a quite common practice and therefore actually very descriptive.

Okay, I didn't know that. I've seen it before somewhere else but didn't think too much of it. Thanks for the info.

> Could you review #3309 and #3175 as well?

Sure, I'll do it later today when I have time.

antis81 commented:
This PR is ready to merge. @zetok Would you do it? I can also merge myself. As this includes some important improvements to the audio code, it would help merging soon.

antis81 commented:
@zetok Bump.

zetok commented:
@antis81 Sorry for no earlier answer.

If PR is ready, LGTM :)
